qiskit/pulse/schedule.py

```diff
@@ -1024,13 +1175,90 @@ def instructions(self) -> Tuple[Tuple[int, Instruction]]:
+        return tuple(blocks)
+
+    def _collect_parameters(
+        self,
+        scope: str,
+        add_scope: bool = False,
+    ) -> Dict[Tuple[str, int], Parameter]:
+        """A helper method to return parameters.
+
+        Parameters defined in the nested subroutines are recursively investigated and
+        returned with parameters within the current scope.
+        To distinguish where the parameters are defined, set ``add_scope=True``.
+
+        Args:
+            scope: Name of current scope.
+            add_scope: Set ``True`` to prepend scope to parameter name.
+
+        Returns:
+            All parameters defined under the current scope.
+        """
+        parameters_out = {}
+        for param in self._parameter_manager.parameters:
+            if add_scope:
+                param = _scope_parameter(param, scope, Reference.scope_delimiter)
+            # This avoids unexpected merging of parameter objects defined in different scopes.
+            # Note that the search_parameters method is scope-aware.
+            # However, when the same parameter object is used in multiple scopes,
+            # if it would return one object from a random scope, the search could fail.
+            # This logic evaluates the parameter equality not only with uuid, but also with name.
+            scope_aware_key = param.name, hash(param)
+            parameters_out[scope_aware_key] = param
+
+        for sub_namespace, subroutine in self.references.items():
+            if subroutine is None:
+                continue
+            composite_key = Reference.key_delimiter.join(sub_namespace)
+            full_path = f"{scope}{Reference.scope_delimiter}{composite_key}"
+            sub_parameters = subroutine._collect_parameters(scope=full_path, add_scope=add_scope)
+            parameters_out.update(sub_parameters)
+
+        return parameters_out
 
     @property
-    def parameters(self) -> Set:
-        """Parameters which determine the schedule behavior."""
-        return self._parameter_manager.parameters
+    def parameters(self) -> Tuple[Parameter]:
+        """Return unassigned parameters with raw names."""
+        return tuple(self._collect_parameters(scope="root", add_scope=False).values())
+
+    @property
+    def scoped_parameters(self) -> Tuple[Parameter]:
+        """Return unassigned parameters with scoped names.
+
+        .. note::
+
+            If a parameter is defined within a nested scope,
+            it is prefixed with all parent-scope names with the delimiter string,
+            which is "::". If a reference key of the scope consists of
+            multiple key strings, it will be represented by a single string joined with ",".
+            For example, "root::xgate,q0::amp" for the parameter "amp" defined in the
+            reference specified by the key strings ("xgate", "q0").
+        """
+        return tuple(self._collect_parameters(scope="root", add_scope=True).values())
```

Comment:
I get nervous for performance reasons when properties do non-trivial work (similar to above, but there's an API concern there as well) - given that this is new, is it worth it being a method?

Reply:
Yes, I converted this into method with some refactoring https://github.com/Qiskit/qiskit-terra/pull/8005/commits/5b7eeef21deec5d01ad0fd542275f135d54d7c55